corehq/ex-submodules/pillowtop/processors/elastic.py

```diff
@@ -67,7 +67,12 @@ def process_change(self, change):
             return
 
         if change.deleted and change.id:
-            self._delete_doc_if_exists(change.id)
+            doc = change.get_document()
+            if doc and doc.get('doc_type') is not None:
+                if doc['doc_type'].endswith("-Deleted"):
+                    self._delete_doc_if_exists(change.id)
```

Comment:
I think there is a race condition here. It would do the wrong thing if the doc is deleted just after `change.get_document()` reads the doc, but before this line is executed, right?

Reply:
Are you describing a situation where the pillow has picked up the change before the original transaction has completed? I don't think that is possible. If you mean another update just happens to be at the same time, that will trigger another message in kafka and when the pillow gets to that one, it will correctly delete the doc. 

I think one behavior change this does introduce is that if a document is deleted and quickly undeleted, there is currently a short period of time where it is not in ES, between when those two changes are processed, and that will not happen anymore. The delete message will just no-op, and then the later undelete will be processed normally, which I think is a slight improvement, but not significantly either way.

Reply:
I think what @millerdev is saying is this sequence of events:

- Change A: doc gets deleted
- Change A gets processed by a pillow which executes through `doc = change.get_document()` which says it's deleted
- At that moment, Change B: doc gets restored
- (x) Change A continues on and deletes the doc
- (y) Change B gets processed and restores the doc

Are we sure that (x) and (y) will happen in the order described above and not in the reverse order (y before x), thus resulting in a deleted doc in the end, rather than a restored doc?

I think the answer "no" is highly accurate without being capital T True. Rather, that the time it takes Change A to get from line 72 to line 76 is going to be orders of magnitude smaller than the time it takes Change B to get from happening (between Change A lines 72 and 76) to being queued, pulled off the queue and getting it its own line 72. Regardless, I think this change takes _existing_ race conditions and reduces them to this much, much narrower sliver. I think the race condition I just outlined is also essentially the same race condition that all changes face, two updates in a row, for instance. Check my logic though

Reply:
Change B never gets to line 72, as this branch only happens if the kafka message is marked as a deletion (see line 71), so the processing of change B, will never accidentally result in a deletion.

That being said, it does look like there might be a race condition, if in between lines 71 and 73 of change A (x), the doc is restored (change B), published to kafka, and processed by the pillow (y). That race condition notably already exists, in a much simpler form, if A and B happen close together and B is processed before A, since we don't guarantee ordering in the pillow. This change makes that race condition impossible, but I guess leaves open the race condition of all of B and subsequent pillow processing happening within these three lines of A. It is hard to believe that is meaningfully possible (it would involve a sql/couch save, publishing to kafka, reading from kafka, fetching the doc from sql/couch, and updating it in ES all happening in the time of a string comparison of a dict value and an es delete), but I can agree that it is theoretically possible (i.e. if the ES delete hangs for an extended period for unknown reasons that do not affect the ES update happening for the other change).

Reply:
Since network calls are involved in the race condition I think it is still possible. Although I agree, this does appear to be an improvement.

At least in Couch I think this is one issue the doc revision (`_rev`) is intended for and can be used to prevent. Is it possible to use the revision when updating the doc in Elastic so that it fails if another actor wins the race to update?

Reply:
Danny and I have talked a bit about this race condition, and I have made a ticket for followup, but since this is still an improvement over the existing situation, I have merged the PR for now.